### DISCUSSION TITLE
Multi Tenancy not working. Need Help

### DISCUSSION THREAD

**Discussion by jaypeemayo** [OrchardCMS/OrchardCore#10749] [Q&A]
Hi  All,

I'm having issues with multi tenancy at the moment.

I have this DNS setting where the main domain is happy.store.com. I have 2 subdomains named adult(adult.happystore.com) and kids(kids.happystore.com). both are pointing to 67.43.15.185 which is the IP of the host site.
![image](https://user-images.githubusercontent.com/21273774/143203543-b4a32050-07cf-458a-a490-03689ee87538.png)

For my orchard core setting:
kids.happystore.com is the default tenant. This one works even if there is no value set.
![image](https://user-images.githubusercontent.com/21273774/143202219-4b1a43dd-b3e7-410f-8bc8-4967605359db.png)

adult.happystore.com is the child tenant. **This one does not work**
![image](https://user-images.githubusercontent.com/21273774/143204291-669df521-a80b-46b2-8a61-63f6f7abac70.png)

Note: also tried to set the hostname to https://adult.happystore.com/ but the browser URL shows https//adult.happystore.com/ (without ":" which is weird). This does not work as well.

We are not sure what we are missing here why this setting does not work. If anyone finds anything wrong or has advice on how to fix please let us know. thanks.

Additional question: is there is an existing issue with multi tenantancy functionality? is this related to this fix? See #10743 @Skrypt @jtkech 

FYI here are my orchardcore versions
OrchardCore.Application.Cms.Core.Targets" Version="1.0.0"
"OrchardCore.Application.Mvc.Targets" Version="1.0.0"

**Comment by Skrypt** [comment]
Please update to latest bits and make sure you add your domain with the currently running ports as an option in the tenant Hostname.
Something like : adult.happystore.com, adult.happystore.com:5001

**Reply by jaypeemayo** [reply]
Thanks for your response @Skrypt ,

Do you mean to update my version to Orchard Core 1.1.0? will this work without the fix in #10743?

**Reply by Skrypt** [reply]
I'm reverting those changes as I think that one of the issues is that it is not properly documented, I needed to understand how it works while debugging it and I never knew we could pass a port with those domain names. This is supposed to work already but in 1.0 there was probably an issue still with the generated URL triggered by the setup and view buttons in the tenant's list page. You need to enter these URLs manually for now but they definitely work.

I've tested some different things and it was intentional that they left it that way so that you can host a tenant on a specific port and a different one on another. So basically your host tenant could be served on port 80 and your child tenants could be served on their own specific port. We don't enforce any specific configuration. Here, the issue is that when you are debugging with VS Code or Visual Studio it is not using port 80 or 443 which are "default" ports for HTTP and HTTPS.

So, you need to specify in that case that you want your tenant to answer on a specific port that you opened through modifying the launchSettings.json file.

```
{
  "iisSettings": {
    "windowsAuthentication": false,
    "anonymousAuthentication": true,
    "iisExpress": {
      "applicationUrl": "http://localhost:2918",
      "sslPort": 44300
    }
  },
  "profiles": {
    "IIS Express": {
      "commandName": "IISExpress",
      "launchBrowser": true,
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      }
    },
    "web": {   // this is Kestrel configuration
      "commandName": "Project",
      "launchBrowser": true,
      "launchUrl": "https://localhost:5001",
      "environmentVariables": {
        "ASPNETCORE_ENVIRONMENT": "Development"
      },
      "applicationUrl": "https://localhost:5001;http://localhost:5000;https://localhost:5003;http://localhost:5002"
    }
  }
}
```

And then in your Tenant configuration in the hostname setting, you add: localhost:5003 and it should answer on https://localhost:5003 for that tenant. There is still an issue with those buttons because the port itself doesn't tell if it uses HTTP or HTTP. So, I will need to talk with @sebastienros about this.

So basically the issue is that if the first domain name added in the HostName tenant setting is specifying a port that is using HTTP and that you are browsing this tenant list page from an HTTPS URL it will append HTTPS on the URL and vice versa.

Now, the mistake that most people do "like me" is that we add only the domain name without a port while we currently need to add all known possibilities. I personally don't like needing to add duplicate domain names with multiple ports in these configs but I think optimally we should be able to define also if the domain port is running through HTTP or HTTPS. So the issue is that the ShellSettings HostNames don't define if a `domain.local:port` runs on HTTP or HTTPS. And then, maybe a solution would be to require to specify it too like : `http\domain.local:port`

**Reply by jaypeemayo** [reply]
Thanks, @Skrypt, sorry a bit confused on how orchard code versioning works. If I update my version 1.1.0, will I get the the version where you reverted the change? Or do I need to go master branch and rebuild everything myself and use those dlls in the meantime?

Appreciate your help. we have been scratching our heads wondering why our setup is not working. We try to include the ports once we update the versions

**Reply by Skrypt** [reply]
It should work in both 1.0.0 and 1.1.0. The issue with the wrong URL though is still current and also in those versions. You will need to guess the proper URL to set up your tenants. I will try to come up with a proper fix in the meantime but I need to discuss this with @sebastienros or @jtkech because we don't want to break any existing tenant at the same time.

**Comment by sebastienros** [comment]
> I'm reverting those changes as I think that one of the issues is that it is not properly documented

I have no idea what you are talking about.

The site can be started with `http://*:80;https://*:443;` to accept any IPs. If you are coming from a proxy you need to enable the forward proxy feature in ASP.NET by setting the correct environment variable, or the initiating domain won't reach your app, and Orchard won't be able to know which domain is targetted.

**Reply by sebastienros** [reply]
Example: https://bartwullems.blogspot.com/2021/03/aspnet-core-aspnetcoreforwardedheaderse.html

**Reply by Skrypt** [reply]
The issue is that the "view" and "setup" buttons don't use these host settings. It uses the first host in that list and it is not making a distinction between HTTP or HTTPS. So, you need to guess the setup tenant URL and enter it manually to get to that page. I didn't try yet with the settings you are suggesting but I will. Also, if this works then maybe the hint underneath the field should mention it. Because there is a lot of people confused right now.

**Reply by Skrypt** [reply]
https://user-images.githubusercontent.com/3228637/143483291-7041eb79-cb3a-4aac-9768-bad3b8ed6df7.mp4

See there what I'm trying and the button doesn't even have a url now.

**Reply by jaypeemayo** [reply]
hmm should the values be comma separated not ";"?

**Comment by sebastienros** [comment]
😢

**Reply by Skrypt** [reply]
😞

**Reply by jaypeemayo** [reply]
@Skrypt 

What I tried:
Update my version to 1.1.0
I asked the hosting to open port number 47871. 

tried the following Hostname values
http://adult.happystore:47871,https://adult.happystore.com:47871
adult.happystore:47871
adult.happystore, adult.happystore:47871
adult.happystore:47871, adult.happystore.com

but no luck to all of the above.

also the ":" is missing 
![image](https://user-images.githubusercontent.com/21273774/143508618-211a09b1-b8c6-478b-84ad-62fd103be39f.png)

Not sure which is having issues now. Is it my hosting's port 47871 or Tenant functionality of OC?

**Reply by Skrypt** [reply]
Have you tried to update the URL manually to http://adult.happystore:47871 ?

**Reply by jtkech** [reply]
Just for info as I didn't follow the discussion ;) Yes, for now the Scheme (http / https) is not an option in the tenant settings, as I remember currently in the tenant settings UI, hosts should be comma separated, and should not  include the Scheme (http / https).

**Reply by jaypeemayo** [reply]
@Skrypt  
tried

For - https://adult.happystore:47871,http://adult.happystore.com:47871
I updated the url to https://adult.happystore:47871

For - http://adult.happystore:47871,https://adult.happystore.com:47871
updated the url to http://adult.happystore:47871

Im getting "This site cant be reached DNS_PROBE_FINISHED_NXDOMAIN" for all combinations that i try.

**Reply by jaypeemayo** [reply]
> Just for info as I didn't follow the discussion ;) Yes, for now the Scheme (http / https) is not an option in the tenant settings, as I remember currently in the tenant settings UI, hosts should be comma separated, and should not include the Scheme (http / https).

![image](https://user-images.githubusercontent.com/21273774/143509986-2142f4fa-de41-4cce-9dd4-c43ff7029419.png)
The help message suggest to include http/https which is confusing.

tried with http/https and without it,no luck for both

**Reply by Skrypt** [reply]
DNS_PROBE_FINISHED_NXDOMAIN this is a DNS issue that has nothing to do with the service working on the port or not. So this is more likely a DNS configuration missing. Or the issue is that you are using a Proxy and using the wrong config along with it.

**Reply by jaypeemayo** [reply]
Also set this and also no luck 
![image](https://user-images.githubusercontent.com/21273774/143623352-2664998c-0972-4aa4-bd0d-84f4b5c2c1c1.png)

**Reply by Skrypt** [reply]
We have a ReverseProxy module that you may want to take a look at. I never tried it yet. Can't help much with this, I would need to try it.

https://github.com/OrchardCMS/OrchardCore/tree/main/src/OrchardCore.Modules/OrchardCore.ReverseProxy

**Comment by jtkech** [comment]
> The help message suggest to include http/https which is confusing.

Yes it says that from a given Host (no Scheme) it generates an URL including the Scheme (for now the one of the current request)

But I agree, may be confusing

**Comment by blueberryfy** [comment]
Folks, please allow me to chime in and contribute. Thanks to the excellent work of our contributors, it was an easy process for me. There might be a better process, but let me share mine that worked:

	1. Working environment: OC 1.7.2, Azure hosting my web app service (B1), godaddy.com for my domain.
	2. My default site is already working with a valid SSL certificate from Azure.
	3. To add a tenant, I initially left the hostname blank but added a url. This is just so I can test the site is working.
	4. Once the tenant site is confirmed working using the url, I went to azure to add the custom domain for the tenant.
	5. I went to godaddy.com to make the necessary changes, as instructed by the azure customer domain process.
	6. I went back to the default OC site to edit the tenant parameters: removed the url and added the hostname.
        7. Gave it around 5 minutes to propagate, then tested.

I came to this thread via the video from Lombiq (thanks, @Piedone). 
https://www.youtube.com/watch?v=2nKvyP6lRWM

The link below was also very helpful:
https://stackoverflow.com/questions/43772971/can-multiple-domains-point-to-the-same-azure-web-app

**Reply by Piedone** [reply]
Awesome!
